rfcs/0032-github-merging-strategy.md

```diff
@@ -0,0 +1,134 @@
+
+We would like for pull-requests (PRs) to merge more quickly and automatically.
+We would also like to encourage smaller PRs, to reduce the burden on reviewers
+and to reduce the amount of work that has to be done keeping up with upstream
+changes in the `develop` branch.
+
+In addition, we have begun to use GitHub's 'PR chains', where a series of
+smaller changes can be linked together in a chain, but with each change in its
+own PR. These allow us to continue to work on larger features even when
+earlier changes haven't yet been included in the `develop` branch, and without
+combining the changes into a single large PR.  PR #5093 added support for the
+`ready-to-merge-into-develop` label, which automates the process of merging
+these PR chains correctly.
+
+When multiple PRs modify the same lines of code and one of them is merged into
+`develop`, the other will have 'merge conflicts' and will need manual
+intervention to fix the conflict before it can be merged. We would like to
+minimise this because
+* PRs with conflicts do not merge automatically
+* the PR process is slowed down by waiting for these manual interventions
+* developer time is wasted in resolving these conflicts.
+
+Our current merge strategy is 'squash and merge', which discards the
+individually committed changes within each PR. This hampers `git`'s ability to
+reduce or eliminate merge conflicts by combining the changes from individual
+commits, unnecessarily creating more merge conflicts to be handled manually.
+
+These unnecessary conflicts are particularly prevelant in PR chains. For
+example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt]
+-> ...
+```
+when `PR 1` is squashed and merged into develop, `git` is unable to recognise
+that the initial changes made in `PR 2` match with those from `PR 1` and
+reports a merge conflict.
+
+If PRs in chains have review comments that result in changes, these will
+usually also result in a merge conflict. For example,
+```
+PR 1
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Modify line 1 in A.txt in response to review comments]
+
+PR 2 (chained to PR 1)
+   develop
+-> [Add line 1 in A.txt]
+-> [Modify line 1 in B.txt]
+-> [Add line 1 in C.txt]
+-> ...
+```
+when `PR 1` is squashed and merged, `git` will again be unable to merge `PR 2`
+automatically.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+We can use `git`'s merge commits when merging to avoid the above issues. These
+preserve the individual commit information that `git` needs to automatically
+resolve these conflicts.
+
+This is a simple configuration change in the GitHub settings for the repo.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+#### We lose a linear `git` history (from `git log` and friends)
+
+We could mitigate this by running a branch in parallel with `develop` that
```

Comment:
Wouldn't this just cause conflicts with other branches which are staged using the `ready-to-merge-into-develop` tags?

Reply:
The idea of this branch is to maintain a history that looks as if we were still using squash, but it should never be used for any actual development. In particular, it shouldn't interact with any PRs (or be PR'd agaist), so there shouldn't be any space for conflicts.

I'm not convinced that there's any inherent value to this alternate history, but it does give us a branch to point people to if they're not willing/able to deal with a non-linear log.